### PR TITLE
Option for changing MongoDB's auth source.

### PR DESCRIPTION
**Pull request by moesoha**

### CODE REVIEW

File: vj4/db.py [hunk 1]
```diff
@@ -22,7 +23,8 @@ def escape(s):
     uri = 'mongodb://{}:{}@{}/?authSource={}'.format(escape(options.db_username),
                                                      escape(options.db_password),
                                                      escape(options.db_host),
-                                                     escape(options.db_name))
+                                                     escape(options.db_auth if options.db_auth
```
Comment (iceb0y) [inline vj4/db.py]: This line is long, so move the others down:

```
    uri = 'mongodb://...'.format(
        escape(...),
        escape(...),
        ...
```

File: vj4/db.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 options.define('db_name', default='test', help='Database name.')
 options.define('db_username', default='', help='Database username.')
 options.define('db_password', default='', help='Database password.')
+options.define('db_auth', default='', help='Database name which stores the user’s credential.')
```
Comment (iceb0y) [inline vj4/db.py]: `’` is not ascii character.

Also see the official description:
https://docs.mongodb.com/manual/reference/connection-string/#urioption.authSource

File: vj4/db.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 options.define('db_name', default='test', help='Database name.')
 options.define('db_username', default='', help='Database username.')
 options.define('db_password', default='', help='Database password.')
+options.define('db_auth', default='', help='Database name which stores the user\'s credential.')
```
Comment (iceb0y) [inline vj4/db.py]: use `db_auth_source`

File: vj4/db.py (hunk 1 above)
Comment (iceb0y) [inline vj4/db.py]: Consider only append `?authSource` when `--db-auth-source` is set, otherwise use `mongodb://u:p@h/name`

Comment (iceb0y) [reply vj4/db.py]: fyi the code to parse this URI string is here:
https://github.com/ZeoAlliance/aiomongo/blob/adeec47b2ef7e2550a797b3202f6604a9da84d67/aiomongo/client.py#L21

**Comment by moesoha** [comment]
@iceb0y done.

File: vj4/db.py [hunk 4]
```diff
@@ -1,29 +1,36 @@
 import aiomongo
 import functools
 import urllib.parse
+from yarl import URL
 
 from vj4.util import options
 
 options.define('db_host', default='localhost', help='Database hostname or IP address.')
+options.define('db_port', default=27017, help='Database hostname or IP address.')
```
Comment (twd2) [inline vj4/db.py]: "Database port."

File: vj4/db.py [hunk 5]
```diff
@@ -1,29 +1,36 @@
...
-                                                     escape(options.db_name))
-  _client = await aiomongo.create_client(uri)
+  url_parts = {
+    'scheme': 'mongodb',
+    'host': options.db_host,
+    'path': options.db_name,
+    'port': options.db_port,
+    'user': options.db_username,
+    'password': options.db_password,
+    'query': {}
```
Comment (twd2) [inline vj4/db.py]: `{'authSource': options.db_auth_source} if options.db_auth_source else {}`?

Comment (moesoha) [reply vj4/db.py]: ↑_↑

File: vj4/db.py [hunk 6]
```diff
@@ -1,29 +1,36 @@
...
+  }
+  if options.db_auth_source:
+    url_parts['query']['authSource'] = options.db_auth_source
+  url = URL.build(**url_parts)
+
+  _client = await aiomongo.create_client(str(url))
   _db = _client.get_database(options.db_name)
```
Comment (iceb0y) [inline vj4/db.py]: you can use `get_default_database()` here since you specified a default database in URL.

File: vj4/db.py [hunk 7]
```diff
@@ -1,29 +1,36 @@
 import aiomongo
 import functools
 import urllib.parse
+from yarl import URL
```
Comment (iceb0y) [inline vj4/db.py]: `import yarl`

The naming convention of this repository doesn't allow importing classes.

File: vj4/db.py [hunk 8]
```diff
@@ -1,29 +1,36 @@
...
-                                                     escape(options.db_name))
-  _client = await aiomongo.create_client(uri)
+  url_parts = {
+    'scheme': 'mongodb',
+    'host': options.db_host,
+    'path': options.db_name,
+    'port': options.db_port,
+    'user': options.db_username,
+    'password': options.db_password,
+    'query': {}
```
Comment (iceb0y) [inline vj4/db.py]: It's nice to have a trailing comma (`,`) here.

**Comment by iceb0y** [comment]
Merging. Thanks!

**Pull request merged**